tmpl/wiki/mirage2-to-mirage3.md

```diff
@@ -0,0 +1,1416 @@
+
+```
+let stack = generic_stackv4 default_network
+```
+
+You may notice that `generic_stackv4` doesn't take a `console` argument, where the previous `direct_stackv4*` and `socket_stackv4` functions did.  This leads us to another big change that predates MirageOS 3, but is more widely used and better supported in this release: logging.
+
+The `mirage_logs` library and its incorporation into the configuration language dates back to version 2.9.0, but a number of libraries still required a `console impl` argument and expected to log to the console.  In Mirage 3.0.0, we've tried to replace the calls to `Console.log` with calls to `Logs.debug`, `Logs.info`, or `Logs.warn` as appropriate.  Consequently, many of the functions which previously needed to be passed a console no longer require it.  For example, we can take [the `stackv4` example from `mirage-skeleton` commit f36d2958f616fb882df37f08d3440797471ca0cc](https://github.com/mirage/mirage-skeleton/tree/f36d2958f616fb882df37f08d3440797471ca0cc/stackv4), which fails with Mirage 3:
+
+```
+mirage-skeleton/stackv4$ cat config.ml
+open Mirage
+
+let handler = foreign "Unikernel.Main" (console @-> stackv4 @-> job)
+
+let stack = generic_stackv4 default_console tap0
+
+let () =
+  register "stackv4" [handler $ default_console $ stack]
+mirage-skeleton/stackv4$ 
+mirage-skeleton/stackv4$ mirage configure -t unix
+mirage: unknown option `-t'.
+Usage: mirage configure [OPTION]... 
+Try `mirage configure --help' or `mirage --help' for more information.
+error while executing ocamlbuild -use-ocamlfind -classic-display -tags
+                        bin_annot -quiet -X _build-ukvm -pkg mirage
+                        config.cmxs
++ ocamlfind ocamlc -c -bin-annot -package mirage -o config.cmo config.ml
+File "config.ml", line 5, characters 44-48:
+Error: The function applied to this argument has type
+         ?group:string ->
+         ?config:Mirage.ipv4_config ->
+         ?dhcp_key:bool Mirage.value ->
+         ?net_key:[ `Direct | `Socket ] Mirage.value ->
+         Mirage.stackv4 Functoria.impl
+This argument cannot be applied without label
+Command exited with code 2.
+Hint: Recursive traversal of subdirectories was not enabled for this build,
+  as the working directory does not look like an ocamlbuild project (no
+  '_tags' or 'myocamlbuild.ml' file). If you have modules in subdirectories,
+  you should add the option "-r" or create an empty '_tags' file.
+  
+  To enable recursive traversal for some subdirectories only, you can use the
+  following '_tags' file:
+  
+      true: -traverse
+      <dir1> or <dir2>: traverse
+```
+
+remove the `default_console` argument to `stackv4`, to give the following `config.ml`:
+
+```ocaml
+open Mirage
+
+let handler = foreign "Unikernel.Main" (console @-> stackv4 @-> job)
+
+let stack = generic_stackv4 tap0
+
+let () =
+  register "stackv4" [handler $ default_console $ stack]
+mirage-skeleton/stackv4$ mirage configure -t unix
+mirage: unknown option `-t'.
```

Comment:
And here. Is this really the new behaviour? We should add an explanation if so (e.g. "note: ignore errors about unknown options if there's also an error about the config.ml file itself").